File: backend/app/schemas/equipment.py
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict
import uuid
from ..models.equipment import EquipmentStatus, LogType, LogOutcome, CleaningStatus


class EquipmentCreate(BaseModel):
    name: str
    category: str
    manufacturer: Optional[str] = None
    model_number: Optional[str] = None
    serial_number: Optional[str] = None
    location: Optional[str] = None
    status: EquipmentStatus = EquipmentStatus.ACTIVE
    last_calibration_date: Optional[date] = None
    calibration_due_date: Optional[date] = None
    calibration_certificate: Optional[str] = None
    last_maintenance_date: Optional[date] = None
    next_maintenance_date: Optional[date] = None
    maintenance_interval_days: Optional[int] = None
    notes: Optional[str] = None


class EquipmentUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    manufacturer: Optional[str] = None
    model_number: Optional[str] = None
    serial_number: Optional[str] = None
    location: Optional[str] = None
    status: Optional[EquipmentStatus] = None
    last_calibration_date: Optional[date] = None
    calibration_due_date: Optional[date] = None
    calibration_certificate: Optional[str] = None
    last_maintenance_date: Optional[date] = None
    next_maintenance_date: Optional[date] = None
    maintenance_interval_days: Optional[int] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class UserBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    full_name: str
    username: str


class EquipmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    equipment_id: str
    name: str
    category: str
    manufacturer: Optional[str]
    model_number: Optional[str]
    serial_number: Optional[str]
    location: Optional[str]
    status: EquipmentStatus
    last_calibration_date: Optional[date]
    calibration_due_date: Optional[date]
    calibration_certificate: Optional[str]
    last_maintenance_date: Optional[date]
    next_maintenance_date: Optional[date]
    maintenance_interval_days: Optional[int]
    notes: Optional[str]
    is_active: bool
    created_at: datetime
    updated_at: datetime
    created_by: Optional[UserBrief]
    cleaning_status: CleaningStatus
    cleaning_status_updated_at: Optional[datetime]


class EquipmentList(BaseModel):
    items: list[EquipmentOut]
    total: int
    page: int
    size: int


class EquipmentLogCreate(BaseModel):
    log_type: LogType
    outcome: LogOutcome = LogOutcome.PASS
    title: str
    description: Optional[str] = None
    performed_by: Optional[str] = None
    performed_date: date
    next_due_date: Optional[date] = None
    certificate_number: Optional[str] = None
    cost: Optional[str] = None
    downtime_hours: Optional[int] = None
    notes: Optional[str] = None


class EquipmentLogUpdate(BaseModel):
    log_type: Optional[LogType] = None
    outcome: Optional[LogOutcome] = None
    title: Optional[str] = None
    description: Optional[str] = None
    performed_by: Optional[str] = None
    performed_date: Optional[date] = None
    next_due_date: Optional[date] = None
    certificate_number: Optional[str] = None
    cost: Optional[str] = None
    downtime_hours: Optional[int] = None
    notes: Optional[str] = None


class EquipmentLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    equipment_id: uuid.UUID
    log_type: LogType
    outcome: LogOutcome
    title: str
    description: Optional[str]
    performed_by: Optional[str]
    performed_date: date
    next_due_date: Optional[date]
    certificate_number: Optional[str]
    cost: Optional[str]
    downtime_hours: Optional[int]
    notes: Optional[str]
    created_at: datetime
    created_by: Optional[UserBrief]


class EquipmentLogList(BaseModel):
    items: list[EquipmentLogOut]
    total: int


class CleaningLogCreate(BaseModel):
    status_to: CleaningStatus
    cleaning_method: Optional[str] = None
    cleaning_agent: Optional[str] = None
    performed_by: Optional[str] = None
    performed_at: Optional[datetime] = None
    verified_by: Optional[str] = None
    verified_at: Optional[datetime] = None
    batch_number: Optional[str] = None
    notes: Optional[str] = None


class CleaningLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    equipment_id: uuid.UUID
    status_from: Optional[CleaningStatus]
    status_to: CleaningStatus
    cleaning_method: Optional[str]
    cleaning_agent: Optional[str]
    performed_by: Optional[str]
    performed_at: datetime
    verified_by: Optional[str]
    verified_at: Optional[datetime]
    batch_number: Optional[str]
    notes: Optional[str]
    created_at: datetime
    created_by: Optional[UserBrief]


class CleaningLogList(BaseModel):
    items: list[CleaningLogOut]
    total: int

File: frontend/src/types/equipment.ts
export type EquipmentStatus =
  | 'ACTIVE'
  | 'INACTIVE'
  | 'UNDER_MAINTENANCE'
  | 'CALIBRATION_DUE'
  | 'RETIRED';

export type CleaningStatus =
  | 'CLEAN'
  | 'DIRTY'
  | 'SANITIZED'
  | 'STERILIZED'
  | 'IN_USE'
  | 'QUARANTINE'
  | 'CLEANING_IN_PROGRESS'
  | 'AWAITING_VERIFICATION';

export type MaterialType =
  | 'API'
  | 'EXCIPIENT'
  | 'PACKAGING'
  | 'SOLVENT'
  | 'REAGENT'
  | 'OTHER';

export type MaterialStatus = 'ACTIVE' | 'INACTIVE' | 'DISCONTINUED';

export interface UserBrief {
  id: string;
  full_name: string;
  username: string;
}

export interface Equipment {
  id: string;
  equipment_id: string;
  name: string;
  category: string;
  manufacturer?: string;
  model_number?: string;
  serial_number?: string;
  location?: string;
  status: EquipmentStatus;
  last_calibration_date?: string;
  calibration_due_date?: string;
  calibration_certificate?: string;
  last_maintenance_date?: string;
  next_maintenance_date?: string;
  maintenance_interval_days?: number;
  notes?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  created_by?: UserBrief;
  cleaning_status: CleaningStatus;
  cleaning_status_updated_at?: string;
}

export interface EquipmentList {
  items: Equipment[];
  total: number;
  page: number;
  size: number;
}

export interface Material {
  id: string;
  material_code: string;
  name: string;
  material_type: MaterialType;
  cas_number?: string;
  pharmacopoeia_standard?: string;
  grade?: string;
  supplier_name?: string;
  supplier_code?: string;
  manufacturer_name?: string;
  unit_of_measure: string;
  storage_conditions?: string;
  shelf_life_days?: number;
  reorder_point?: string;
  notes?: string;
  status: MaterialStatus;
  created_at: string;
  updated_at: string;
  created_by?: UserBrief;
}

export interface MaterialList {
  items: Material[];
  total: number;
  page: number;
  size: number;
}
